README.rst
Stepper Brick
=============

This repository contains the firmware source code and the hardware design
files. The documentation generator configs can be found at
https://github.com/Tinkerforge/generators

Repository Content
------------------

software/:
 * examples/: Examples for all supported languages
 * build/: Compiled files
 * src/: Source code of firmware
 * Makefile: Makefile to build project

hardware/:
 * Contains KiCad project files and additionally schematics as PDF

datasheets/:
 * Contains datasheets for sensors and complex ICs that are used

Hardware
--------

The hardware is designed with the open source EDA Suite KiCad
(http://www.kicad-pcb.org). Before you are able to open the files,
you have to install the Tinkerforge kicad-libraries
(https://github.com/Tinkerforge/kicad-libraries). You can either clone
them directly in hardware/ or clone them in a separate folder and
symlink them into hardware/
(ln -s kicad_path/kicad-libraries project_path/hardware). After that you
can open the .pro file in hardware/ with KiCad and from there view and
modify the schematics and the PCB layout.

Software
--------

If you want to do your own Brick/Bricklet firmware development we highly
recommend that you use our build environment setup script and read the
tutorial (https://www.tinkerforge.com/en/doc/Tutorials/Tutorial_Build_Environment/Tutorial.html).

To compile the C code we recommend you to install the newest GNU Arm Embedded
Toolchain (https://launchpad.net/gcc-arm-embedded/+download).
You also need to install bricklib (https://github.com/Tinkerforge/bricklib).

You can either clone it directly in software/src/ or clone it in a
separate folder and symlink it into software/src/
(ln -s bricklib_path/bricklib project_path/software/src/). Finally make sure to
have CMake installed (http://www.cmake.org/cmake/resources/software.html).

After that you can build the firmware by invoking make in software/.
The firmware (.bin) can then be found in software/build/ and uploaded
with brickv (click button "Flashing" on start screen).


software/changelog
2011-12-15: 1.0.0 (35935ef)
- Initial version

2012-01-02: 1.0.1 (01d2743)
- Add code for voltage divider to allow motor current below 100mA

2012-02-15: 1.1.0 (ef71419)
- Add support for tick types

2012-03-05: 1.1.1 (d898cbf)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.1.2 (2539276)
- Fix bug with ADC calibration readout (through bricklib)

2012-04-03: 1.1.3 (0d5a8e3)
- Add averaging for current measurement

2012-04-27: 1.1.4 (6a8b43f)
- Add get_chip_temperature and reset function

2012-07-02: 1.1.5 (aa1e9e2)
- Add USB hotplug support (through bricklib)

2012-07-03: 1.1.6 (f510f97)
- Add time base API, add all-data function and callback, add state callback

2012-07-24: 1.1.7 (74637b6)
- Stop state machine after disable call, don't drive if max velocity = 0

2012-07-27: 1.1.8 (d41a6e4)
- Stop completely if max velocity is set to 0 while in drive mode

2012-10-26: 1.1.9 (0990448)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2013-01-17: 2.0.0 (c53a3b5)
- Support for Protocol 2.0

2013-01-23: 2.0.1 (456df76)
- Add missing header to minimum voltage setter

2013-04-11: 2.0.2 (019d2cc)
- Add watchdog timer (restart after 16s of "deadlock")

2013-05-28: 2.0.3 (6af702f)
- Fix bug in under voltage callback definition (wrong FID)

2013-06-27: 2.0.4 (ce91778)
- Use tick_task to detect USB connect/disconnect instead of interrupt

2014-07-24: 2.0.5 (7ec8e10)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (4f88c2f)
- Add support for new SPI protocol

2015-05-19: 2.3.1 (7967f4b)
- Add magic number for unflashed Bricklet detection (through bricklib)
- Update state if only one step is set

2015-09-17: 2.3.2 (be8e75f)
- Increase external detection voltage to 6V

2016-08-29: 2.3.3 (6ad7c40)
- Make sure to use all chunks when writing Bricklet plugin (through bricklib)

2017-05-11: 2.3.4 (a68497b)
- Add support for Co-MCU Bricklets

2017-06-28: 2.3.5 (71ce9db)
- Fix incompatibility between Co-MCU support and LED Strip Bricklet (through bricklib)

2017-12-13: 2.3.6 (e4a4953)
- Make sure Brick does not restart on suspend/resume (through bricklib)
- Refactor initial enumeration handling, fix double-enumeration bugs (through bricklib)

2018-01-11: 2.3.7 (ac2fdd6)
- Fix message duplication bug (through bricklib)

2018-04-27: 2.3.8 (ab07a14)
- Fix potential deadlock between USB stack and Co-MCU enumeration (through bricklib)

2018-07-31: 2.3.9 (036986c)
- Add support for Isolator Bricklet (through bricklib)
- Improve Co-MCU Bricklet error handling (through bricklib)

2019-02-27: 2.3.10 (ee6d538)
- Changes through bricklib:
  - Fix Bricklet port miss-identification
  - Increase SPI stack timeout (for RED Brick)
  - Fix bugs in set/get_spitfp_* functions


software/bindings/csharp/BrickStepper.cs
/*************************************************************
 * This file was automatically generated on 2011-09-20.      *
 *                                                           *
 * If you have a bugfix for this file and want to commit it, *
 * please fix the bug in the generator. You can find a link  *
 * to the generator git on tinkerforge.com                   *
 *************************************************************/

namespace Tinkerforge
{
	public class BrickStepper : Device 
	{
		private static byte TYPE_SET_MAX_VELOCITY = 1;
		private static byte TYPE_GET_MAX_VELOCITY = 2;
		private static byte TYPE_GET_CURRENT_VELOCITY = 3;
		private static byte TYPE_SET_SPEED_RAMPING = 4;
		private static byte TYPE_GET_SPEED_RAMPING = 5;
		private static byte TYPE_FULL_BRAKE = 6;
		private static byte TYPE_SET_CURRENT_POSITION = 7;
		private static byte TYPE_GET_CURRENT_POSITION = 8;
		private static byte TYPE_SET_TARGET_POSITION = 9;
		private static byte TYPE_GET_TARGET_POSITION = 10;
		private static byte TYPE_SET_STEPS = 11;
		private static byte TYPE_GET_STEPS = 12;
		private static byte TYPE_GET_REMAINING_STEPS = 13;
		private static byte TYPE_SET_STEP_MODE = 14;
		private static byte TYPE_GET_STEP_MODE = 15;
		private static byte TYPE_DRIVE_FORWARD = 16;
		private static byte TYPE_DRIVE_BACKWARD = 17;
		private static byte TYPE_STOP = 18;
		private static byte TYPE_GET_STACK_INPUT_VOLTAGE = 19;
		private static byte TYPE_GET_EXTERNAL_INPUT_VOLTAGE = 20;
		private static byte TYPE_GET_CURRENT_CONSUMPTION = 21;
		private static byte TYPE_SET_MOTOR_CURRENT = 22;
		private static byte TYPE_GET_MOTOR_CURRENT = 23;
		private static byte TYPE_ENABLE = 24;
		private static byte TYPE_DISABLE = 25;
		private static byte TYPE_IS_ENABLED = 26;
		private static byte TYPE_SET_DECAY = 27;
		private static byte TYPE_GET_DECAY = 28;
		private static byte TYPE_SET_MINIMUM_VOLTAGE = 29;
		private static byte TYPE_GET_MINIMUM_VOLTAGE = 30;
		private static byte TYPE_UNDER_VOLTAGE = 31;
		private static byte TYPE_POSITION_REACHED = 32;

		public delegate void UnderVoltage(ushort voltage);
		public delegate void PositionReached(int position);

		public BrickStepper(string uid) : base(uid) 
		{
			messageCallbacks[TYPE_UNDER_VOLTAGE] = new MessageCallback(CallbackUnderVoltage);
			messageCallbacks[TYPE_POSITION_REACHED] = new MessageCallback(CallbackPositionReached);
		}

		public void SetMaxVelocity(ushort velocity)
		{
			byte[] data = new byte[6];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_MAX_VELOCITY, 1, data);
			LEConverter.To((ushort)6, 2, data);
			LEConverter.To(velocity, 4, data);

			ipcon.Write(this, data, TYPE_SET_MAX_VELOCITY, false);
		}

		public void GetMaxVelocity(out ushort velocity)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_MAX_VELOCITY, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_MAX_VELOCITY, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetMaxVelocity in time");
			}

			velocity = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void GetCurrentVelocity(out ushort velocity)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_CURRENT_VELOCITY, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_CURRENT_VELOCITY, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetCurrentVelocity in time");
			}

			velocity = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void SetSpeedRamping(ushort acceleration, ushort deacceleration)
		{
			byte[] data = new byte[8];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_SPEED_RAMPING, 1, data);
			LEConverter.To((ushort)8, 2, data);
			LEConverter.To(acceleration, 4, data);
			LEConverter.To(deacceleration, 6, data);

			ipcon.Write(this, data, TYPE_SET_SPEED_RAMPING, false);
		}

		public void GetSpeedRamping(out ushort acceleration, out ushort deacceleration)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_SPEED_RAMPING, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_SPEED_RAMPING, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetSpeedRamping in time");
			}

			acceleration = LEConverter.UShortFrom(4, answer);
			deacceleration = LEConverter.UShortFrom(6, answer);

			writeEvent.Set();
		}

		public void FullBrake()
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_FULL_BRAKE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_FULL_BRAKE, false);
		}

		public void SetCurrentPosition(int position)
		{
			byte[] data = new byte[8];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_CURRENT_POSITION, 1, data);
			LEConverter.To((ushort)8, 2, data);
			LEConverter.To(position, 4, data);

			ipcon.Write(this, data, TYPE_SET_CURRENT_POSITION, false);
		}

		public void GetCurrentPosition(out int position)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_CURRENT_POSITION, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_CURRENT_POSITION, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetCurrentPosition in time");
			}

			position = LEConverter.IntFrom(4, answer);

			writeEvent.Set();
		}

		public void SetTargetPosition(int position)
		{
			byte[] data = new byte[8];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_TARGET_POSITION, 1, data);
			LEConverter.To((ushort)8, 2, data);
			LEConverter.To(position, 4, data);

			ipcon.Write(this, data, TYPE_SET_TARGET_POSITION, false);
		}

		public void GetTargetPosition(out int position)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_TARGET_POSITION, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_TARGET_POSITION, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetTargetPosition in time");
			}

			position = LEConverter.IntFrom(4, answer);

			writeEvent.Set();
		}

		public void SetSteps(int steps)
		{
			byte[] data = new byte[8];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_STEPS, 1, data);
			LEConverter.To((ushort)8, 2, data);
			LEConverter.To(steps, 4, data);

			ipcon.Write(this, data, TYPE_SET_STEPS, false);
		}

		public void GetSteps(out int steps)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_STEPS, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_STEPS, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetSteps in time");
			}

			steps = LEConverter.IntFrom(4, answer);

			writeEvent.Set();
		}

		public void GetRemainingSteps(out int steps)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_REMAINING_STEPS, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_REMAINING_STEPS, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetRemainingSteps in time");
			}

			steps = LEConverter.IntFrom(4, answer);

			writeEvent.Set();
		}

		public void SetStepMode(byte mode)
		{
			byte[] data = new byte[5];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_STEP_MODE, 1, data);
			LEConverter.To((ushort)5, 2, data);
			LEConverter.To(mode, 4, data);

			ipcon.Write(this, data, TYPE_SET_STEP_MODE, false);
		}

		public void GetStepMode(out byte mode)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_STEP_MODE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_STEP_MODE, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetStepMode in time");
			}

			mode = LEConverter.ByteFrom(4, answer);

			writeEvent.Set();
		}

		public void DriveForward()
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_DRIVE_FORWARD, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_DRIVE_FORWARD, false);
		}

		public void DriveBackward()
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_DRIVE_BACKWARD, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_DRIVE_BACKWARD, false);
		}

		public void Stop()
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_STOP, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_STOP, false);
		}

		public void GetStackInputVoltage(out ushort voltage)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_STACK_INPUT_VOLTAGE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_STACK_INPUT_VOLTAGE, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetStackInputVoltage in time");
			}

			voltage = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void GetExternalInputVoltage(out ushort voltage)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_EXTERNAL_INPUT_VOLTAGE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_EXTERNAL_INPUT_VOLTAGE, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetExternalInputVoltage in time");
			}

			voltage = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void GetCurrentConsumption(out ushort current)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_CURRENT_CONSUMPTION, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_CURRENT_CONSUMPTION, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetCurrentConsumption in time");
			}

			current = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void SetMotorCurrent(ushort current)
		{
			byte[] data = new byte[6];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_MOTOR_CURRENT, 1, data);
			LEConverter.To((ushort)6, 2, data);
			LEConverter.To(current, 4, data);

			ipcon.Write(this, data, TYPE_SET_MOTOR_CURRENT, false);
		}

		public void GetMotorCurrent(out ushort current)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_MOTOR_CURRENT, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_MOTOR_CURRENT, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetMotorCurrent in time");
			}

			current = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void Enable()
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_ENABLE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_ENABLE, false);
		}

		public void Disable()
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_DISABLE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_DISABLE, false);
		}

		public void IsEnabled(out bool enabled)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_IS_ENABLED, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_IS_ENABLED, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for IsEnabled in time");
			}

			enabled = LEConverter.BoolFrom(4, answer);

			writeEvent.Set();
		}

		public void SetDecay(ushort decay)
		{
			byte[] data = new byte[6];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_DECAY, 1, data);
			LEConverter.To((ushort)6, 2, data);
			LEConverter.To(decay, 4, data);

			ipcon.Write(this, data, TYPE_SET_DECAY, false);
		}

		public void GetDecay(out ushort decay)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_DECAY, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_DECAY, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetDecay in time");
			}

			decay = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public void SetMinimumVoltage(ushort voltage)
		{
			byte[] data = new byte[6];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_SET_MINIMUM_VOLTAGE, 1, data);
			LEConverter.To((ushort)6, 2, data);
			LEConverter.To(voltage, 4, data);

			ipcon.Write(this, data, TYPE_SET_MINIMUM_VOLTAGE, false);
		}

		public void GetMinimumVoltage(out ushort voltage)
		{
			byte[] data = new byte[4];
			LEConverter.To(stackID, 0, data);
			LEConverter.To(TYPE_GET_MINIMUM_VOLTAGE, 1, data);
			LEConverter.To((ushort)4, 2, data);

			ipcon.Write(this, data, TYPE_GET_MINIMUM_VOLTAGE, true);

			byte[] answer;
			if(!answerQueue.TryDequeue(out answer, IPConnection.TIMEOUT_ANSWER))
			{
				throw new TimeoutException("Did not receive answer for GetMinimumVoltage in time");
			}

			voltage = LEConverter.UShortFrom(4, answer);

			writeEvent.Set();
		}

		public int CallbackUnderVoltage(byte[] data)
		{
			ushort voltage = LEConverter.UShortFrom(4, data);

			((UnderVoltage)callbacks[TYPE_UNDER_VOLTAGE])(voltage);
			return 6;
		}

		public int CallbackPositionReached(byte[] data)
		{
			int position = LEConverter.IntFrom(4, data);

			((PositionReached)callbacks[TYPE_POSITION_REACHED])(position);
			return 8;
		}

		public void RegisterCallback(System.Delegate d)
		{
			if(d.GetType() == typeof(UnderVoltage))
			{
				callbacks[TYPE_UNDER_VOLTAGE] = d;
			}
			else if(d.GetType() == typeof(PositionReached))
			{
				callbacks[TYPE_POSITION_REACHED] = d;
			}
		}
	}
}
